orca/call_metric_recorder.go

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package orca
+
+import "context"
+
+type callMetricReporterCtxKey struct{}
+
+// GetCallMetricRecorder returns the RPC specific custom metrics recorder
+// [CallMetricRecorder] embedded in the provided RPC context.
+func GetCallMetricRecorder(ctx context.Context) (r *CallMetricRecorder, ok bool) {
+	r, ok = ctx.Value(callMetricReporterCtxKey{}).(*CallMetricRecorder)
+	return
+}
+
+func setCallMetricRecorder(ctx context.Context, r *CallMetricRecorder) context.Context {
+	return context.WithValue(ctx, callMetricReporterCtxKey{}, r)
+}
+
+// CallMetricRecorder supports injection of per-RPC custom backend metrics from
+// the server application.
+//
+// An instance of this type is created for every RPC when custom backend metrics
+// reporting is enabled. A reference to the created instance can be retrieved
+// by the server application using a call to GetCallMetricRecorder().
+//
+// Recording the same metric multiple times overrides the previously recorded
+// values. The methods can be called at any time during an RPC lifecycle.
+//
+// Safe for concurrent use.
+type CallMetricRecorder struct {
```

Comment:
I'm using `MetricSetter` instead of `CallMetricRecorder` since the former also needs to be embedded in the `Service`. I updated all other functions which accept or return a `CallMetricRecorder`.

Reply:
> I'm using `MetricSetter` instead of `CallMetricRecorder` since the former also needs to be embedded in the `Service`. I updated all other functions which accept or return a `CallMetricRecorder`.

Java uses the term `CallMetricRecorder` (and just `MetricRecorder`).  It would be nice to use the same names (not required; just nice).  We _could_ create two interfaces with these names, both of which are satisfied by `metricRecorder`, and have overlapping methods; that seems okay to me.

Reply:
Are you suggesting we have two interfaces:

```golang
type CallMetricRecorder interface { // Has the 4 methods required for call metrics }

type OutOfBandMetricRecorder interface { // Has all the methods required for oob metrics }
```

And we would have the same way to retrieve a reference to the per-call metric recorder:
```golang
func CallMetricReporterFromContext(ctx context.Context) CallMetricRecorder { ... }
```

and the `Service` implementation would embed `OutOfBandMetricRecorder`.

The more we dabble with these interfaces, the less I'm convinced about them. Why don't we revert to simply having concrete types which forward the corresponding calls to an unexported `metricRecorder`. With interfaces, we have the issue of not being able to change the API easily.

Do you see any other benefits of defining these interfaces other than just getting rid of some boiler plate method implementations which delegate work to the underlying `metricRecorder`?

Reply:
I'd really like to see two symbols exported: `orca.CallMetricRecorder` and `orca.MetricRecorder` to correspond to Java's implementation.  Whether they are interfaces or structs matters less to me.  We can document the interfaces and say they are not to be implemented by external packages, and add a private method to them to prevent such use.

Related to my other comment, though, it might not be a good idea to have the same struct behind them if we did use interfaces, anyway, as the per-call metric recorder should ideally be more lightweight given its usage patterns.  So maybe two structs is best, but _not_ delegating to the same backing struct.

Reply:
Switched to having two concrete struct implementations.